### ISSUE TITLE
secLevelLabels with unique levels?

### ISSUE THREAD

**Issue by ricopicone** [lierdakil/pandoc-crossref#295]
My "numbering" scheme is one of abbreviation: each chapter gets a unique abbreviation like "intro" and similarly each section (e.g. "dyn"). This is motivated by my constant reshuffling of material and therefore numbers/references changing.

I have implemented this in LaTeX natively, and as I move over to Pandoc Markdown source, I think I'll be able to find a way to connect my LaTeX style to the output.

But for HTML output at least, I'm hoping to use pandoc-crossref. I was able to get close using `secLevelLabels` like the following.

```yaml
chapters: true
chaptersDepth: 2 # trick it into labeling by section
secLevelLabels:
  - ['']
  - ['sec1','sec2','sec3'] # eg
```

So I can consecutively define the section "numbers"/abbreviations ... this gives equations labels like `(.sec1.1)`. It would be nice to get rid of the leading `.`.

Ideally, though, I could define a unique section abbreviation for each unique chapter abbreviation. Essentially, I think this syntax would be how it could work.

```yaml
chapters: true
chaptersDepth: 2 # trick it into labeling by section
secLevelLabels:
  - ['ch1','ch2','ch3'] # etc
  - [
    ['sec1','sec2','sec3'], # for ch1
    ['foo1','foo2','foo3'], # for ch2
    ['bar1','bar2','bar3','bar4'] # for ch3
  ]
```

I'm not sure if nested arrays or dictionaries are better for section abbreviations. Anyhow, I'm guessing it wouldn't be terribly hard to extend this `secLevelLabels` option as I've described, but I don't know Haskell ... yet, anyhow!

(This is an excellent project, btw&mdash;great work!)

**Comment by ricopicone** [comment]
I'll mention that I'm happy to try to make this change/feature in a fork but would appreciate some guidance on where to make the change. I'm thinking probably `CustomLabels.hs`?

**Comment by ricopicone** [comment]
A further note: my initial "workaround" doesn't really work ... every chapter starts the section numbers over, so I can't really consecutively number them throughout the entire document. That means I'll have to hope for an implementation of my suggestion. Thanks for considering it!

**Comment by lierdakil** [comment]
Uh... sorry, I missed this somehow. So... yeah, I can't really say I understand what exactly you're trying to achieve here, but you could kinda get it to work with a little preprocessing I think? That is, if I understand what you're trying to do.

So, first of all, you can override label for any element, i.e. something like this

```markdown
---
chapDelim: ''
chapters: true
chaptersDepth: 2
sectionsDepth: -1
numberSections: true
---

# Intro {#sec:intro label=""}

## Subsection {#sec:subs label="Intro Subsection."}

[$$x^2=y^3$$]{#eq:eq1 label="eqn"}
```
should be equivalent to 
```markdown
#  Intro

## Intro Subsection. Subsection

$$x^2=y^3\qquad(Intro Subsection.eqn)$$
```

Second, you could probably make a Lua filter that would set section labels for you, and run it before pandoc-crossref.

There are some caveats however. For one, if you want to reference a section directly, you'll have to live with a stray dot, that is,
```markdown
As written in @sec:subs, ...
```
will turn into
```markdown
As written in sec. Intro Subsection., ...
```
which is likely not what you want.

That said, the only reason why I set `chapDelim` to empty string and explicitly added the delimiter to the section label is because empty labels aren't handled gracefully. There is another bug that needs fixing for all of this to work anyway (fixed in master, you can grab binaries from https://github.com/lierdakil/pandoc-crossref/actions/runs/602586964 once it builds), so I can actually "fix" handling of empty labels (by filtering them out) before next release. But I want to make sure it would actually solve your issue first.

I'm open to suggestions, however your use case is _very_ niche, and if we could figure out a reasonable workaround, I'd prefer to avoid adding yet another special case to the code base. Too much of that already, which can be pretty hard to maintain.

**Comment by lierdakil** [comment]
Actually, scratch that. This should probably work: https://github.com/lierdakil/pandoc-crossref/actions/runs/602685626

Use explicit `label` attribute. Set empty `label` on top-level headers if you need to. The last one you can do with a lua filter if you want:
```lua
function Header(el)
    if el.level == 1 then
      el.attributes.label=""
      return el
    end
end
```

Consider:

```markdown
---
chapters: true
chaptersDepth: 2
eqnPrefixTemplate: ($$i$$)
---

# Intro {label=""}

## Subsection {#sec:subs label="intro-sub"}

[$$x^2=y^3$$]{#eq:eq1}

@sec:subs

@eq:eq1
```
turns into
```markdown
# Intro {#intro label=""}

## Subsection {#sec:subs label="intro-sub"}

[$$x^2=y^3\qquad(intro-sub.1)$$]{#eq:eq1}

sec. intro-sub

(intro-sub.1)
```

**Comment by ricopicone** [comment]
@lierdakil this works perfectly! It's great for my a-sequential numbering/labeling system ... thanks!

**Comment by lierdakil** [comment]
Released in v0.3.10.0
